### ISSUE TITLE
OpacityEffect & ColorEffect not working on TextComponents with HasPaint mixin

### ISSUE THREAD

**Issue by jeevanson-vivriti** [flame-engine/flame#1013]
# Current bug behaviour

OpacityEffect is not working on TextComponent

# Expected behaviour

It should work as expected when used with component that has mixin HasPaint

# Steps to reproduce

import 'package:flame/components.dart';
import 'package:flame/effects.dart';
import 'package:flame/game.dart';
import 'package:flame/input.dart';
import 'package:flutter/material.dart';

class OpacityEffectGame extends FlameGame {
  SpriteComponent sprite;

  @override
  Future<void> onLoad() async {
    await super.onLoad();
    final textConfig =
    TextPaint(config: TextPaintConfig(color: Colors.white, fontSize: 35));
    final newStageComponent = TextEffectComponent('Stage',
        config: textConfig, position: Vector2(size.x/2, size.y / 2))
      ..anchor = Anchor.center;
      await newStageComponent.add(
          OpacityEffect(
            opacity: 0,
            duration: 0.5,
            isInfinite: true,
            isAlternating: true,
          ));
    await add(newStageComponent);
    
  }
  
}

class TextEffectComponent extends TextComponent with HasPaint {
  TextEffectComponent(String text,{Vector2 position,TextRenderer config}) : super(text, position: position, textRenderer: config);
}

# Flutter doctor output

```
Output of: flutter doctor -v
```
 Flutter (Channel stable, 2.5.2, on macOS 11.6 20G165 darwin-x64, locale en-IN)
    • Flutter version 2.5.2 at /Users/jeevanson.m/Documents/Work/Flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3595343e20 (10 days ago), 2021-09-30 12:58:18 -0700
    • Engine revision 6ac856380f
    • Dart version 2.14.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/jeevanson.m/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 13.0, Build version 13A233
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] Connected device (2 available)
    • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 94.0.4606.71

• No issues found!

# More environment information

Flame version 
  flame: ^1.0.0-releasecandidate.14

# Log information

```
Enter log information in this code block
```

# More information

**Comment by erickzanardo** [comment]
Unfortunately TextComponent don't support those effects. Marking and enhancement.

**Comment by tty215** [comment]
Is there a workaround for this problem?

**Comment by shtse8** [comment]
> Is there a workaround for this problem?

I have made a custom `TextComponentEx` to extend `TextComponent` for workaround.

```dart

class TextComponentEx extends TextComponent with HasPaint {
  TextComponentEx(
      {super.anchor,
      super.angle,
      super.children,
      super.position,
      super.priority,
      super.scale,
      super.size,
      super.text,
      super.textRenderer});

  // @override
  // Future<void> onLoad() async {
  //   await super.onLoad();

  //   if (textRenderer is TextPaint) {
  //     var foreground = (textRenderer as TextPaint).style.foreground;
  //     if (foreground != null) {
  //       setPaint(this, foreground);
  //     }
  //   }
  // }

  double lastOpacity = 255;
  @override
  void update(double dt) {
    super.update(dt);
    if (opacity == lastOpacity) return;
    if (textRenderer is TextPaint) {
      lastOpacity = opacity;
      textRenderer = ((textRenderer as TextPaint).copyWith(((style) {
        var foreground = style.foreground;
        if (foreground != null) {
          foreground.color = foreground.color.withOpacity(opacity);
        }

        return style.copyWith(
            color: style.color?.withOpacity(opacity),
            foreground: foreground,
            shadows: style.shadows
                ?.map((e) => Shadow(
                    blurRadius: e.blurRadius,
                    color: e.color.withOpacity(opacity),
                    offset: e.offset))
                .toList());
      })));
    }
  }
}
```

**Comment by spydon** [comment]
> I have made a custom `TextComponentEx` to extend `TextComponent` for workaround. Just a workaround, performance is not a consideration.

Do note that this workaround is very resource intense since it creates a lot of new objects in each tick, so you might experience a drop in FPS if you use several components like this.

**Comment by shtse8** [comment]
> > I have made a custom `TextComponentEx` to extend `TextComponent` for workaround. Just a workaround, performance is not a consideration.
> 
> Do note that this workaround is very resource intense since it creates a lot of new objects in each tick, so you might experience a drop in FPS if you use several components like this.

yeah. the performance is bad if there are several hundreds of text components fading out.

This is my second try.

```dart
class TextComponentEx extends PositionComponent with HasPaint {
  TextStyle style;
  TextDirection textDirection;
  String text;
  late TextPainter _painter;
  TextComponentEx(
      {required this.text,
      super.anchor,
      this.textDirection = TextDirection.ltr,
      required this.style}) {
    _painter = TextPainter(
        textDirection: textDirection,
        text: TextSpan(
          text: text,
          style: style,
        ));

    _painter.layout();
  }

  @override
  void render(Canvas canvas) {
    super.render(canvas);
    canvas.saveLayer(
        _painter.size.toRect(), Paint()..blendMode = BlendMode.srcOver);
    _painter.paint(canvas, Offset.zero);
    canvas.drawPaint(Paint()
      ..blendMode = BlendMode.modulate
      ..color = Colors.white.withOpacity(opacity));
    canvas.restore();
  }
}
```

**Comment by spydon** [comment]
@shtse8 you shouldn't have to take the position into consideration in the render method, the canvas should already be positioned according to `position` when it runs.

You can also move out the creation of the `Paint` object to outside of the render method since it is not changing.

**Comment by spydon** [comment]
Something like this will be a bit more efficient since it doesn't have to recreate the expensive `Paint` objects each tick:

```dart
mixin HasOpacityProvider on Component implements OpacityProvider {
  double _opacity = 1;
  final Paint _paint = BasicPalette.white.paint();
  final Paint _srcOverPaint = Paint()..blendMode = BlendMode.srcOver);

  @override
  double get opacity => _opacity;

  @override
  set opacity(double value) {
    if (value == _opacity) return;
    _opacity = value;
    _paint
      ..color = Colors.white.withOpacity(value)
      ..blendMode = BlendMode.modulate;
  }

  @override
  void renderTree(Canvas canvas) {
    canvas.saveLayer(null, _srcOverPaint);
    super.renderTree(canvas);
    canvas.drawPaint(_paint);
    canvas.restore();
  }
}
```
